### PR TITLE
Add SCTP support

### LINKED ISSUE #5719: Add support for SCTP protocol
Introduced in k8s 1.12

https://github.com/kubernetes/kubernetes/pull/64973

Tasks:
* [x] Introspect SCTP headers in datapath
* [x] Support Hubble observability for SCTP
* [x] Support policy enforcement on SCTP packets
* [ ] Support LB / Service translation for SCTP packets

### LINKED ISSUE #6907: SCTP protocol is not working in K8S 1.13.2 and cilium cilium-1.4.0-rc1
## Support

Hi,

I installed K8S version 1.13.2 and Cilium "cilium-1.4.0-rc1" on Ubuntu 18.04.
K8S was installed using Kubeadm

I created new service for SCTP & TCP protocols using heml chart, all processes are up and running, the TCP is working fine but it seems to be that SCTP is not working.
according to the documentation, it should be supported on cilium version 1.4.

**pods and services status:**
 NAMESPACE     NAME                                                  READY   STATUS             RESTARTS   AGE     IP              NODE           NOMINATED NODE   READINESS GATES
cu-sim        pod/cu-sim-cu-sim-deployment-7464ff6bd4-c8xl2         1/1     Running            769        3d22h   10.244.1.215    k8sadmnode     <none>           <none>
kube-system   pod/cilium-bbjxv                                      1/1     Running            0          3d23h   135.25.86.134   k8sadmnode     <none>           <none>
kube-system   pod/cilium-etcd-b6dhxs2mnt                            1/1     Running            0          22h     10.244.1.74     k8sadmnode     <none>           <none>
kube-system   pod/cilium-etcd-cgbqxhzqqr                            1/1     Running            0          3d23h   10.244.1.52     k8sadmnode     <none>           <none>
kube-system   pod/cilium-etcd-mrkfdwhwph                            1/1     Running            0          3d23h   10.244.1.199    k8sadmnode     <none>           <none>
kube-system   pod/cilium-etcd-operator-6c64f47c69-fnv5c             1/1     Running            0          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/cilium-mq7mv                                      1/1     Running            1          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/cilium-operator-576d67d4c4-777nm                  1/1     Running            0          3d23h   10.244.1.38     k8sadmnode     <none>           <none>
kube-system   pod/coredns-86c58d9df4-p92f2                          1/1     Running            0          3d23h   10.244.0.79     k8sadmmaster   <none>           <none>
kube-system   pod/coredns-86c58d9df4-pl5wj                          1/1     Running            0          3d23h   10.244.0.143    k8sadmmaster   <none>           <none>
kube-system   pod/etcd-k8sadmmaster                                 1/1     Running            0          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/etcd-operator-5cf67779fd-nl9tc                    1/1     Running            0          3d23h   10.244.1.23     k8sadmnode     <none>           <none>
kube-system   pod/kube-apiserver-k8sadmmaster                       1/1     Running            0          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/kube-controller-manager-k8sadmmaster              1/1     Running            0          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/kube-proxy-ql5xm                                  1/1     Running            0          3d23h   135.25.86.134   k8sadmnode     <none>           <none>
kube-system   pod/kube-proxy-tfdzk                                  1/1     Running            0          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/kube-scheduler-k8sadmmaster                       1/1     Running            0          3d23h   135.25.86.133   k8sadmmaster   <none>           <none>
kube-system   pod/tiller-deploy-c8bfd74c4-6lns2                     1/1     Running            0          3d22h   10.244.1.60     k8sadmnode     <none>           <none>
rapp-demo     pod/rapp-demo-rapp-demo-deployment-5888bfc84d-f6vw5   0/1     CrashLoopBackOff   830        2d22h   10.244.1.53     k8sadmnode     <none>           <none>
ric-e2        pod/ric-e2-ric-e2-deployment-6b486d85bd-2jv2x         1/1     Running            0          2d23h   10.244.1.130    k8sadmnode     <none>           <none>

NAMESPACE     NAME                         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                          AGE     SELECTOR
default       service/kubernetes           ClusterIP   10.96.0.1        <none>        443/TCP                                          3d23h   <none>
kube-system   service/cilium-etcd          ClusterIP   None             <none>        2379/TCP,2380/TCP                                3d23h   app=etcd,etcd_cluster=cilium-etcd
kube-system   service/cilium-etcd-client   ClusterIP   10.108.48.103    <none>        2379/TCP                                         3d23h   app=etcd,etcd_cluster=cilium-etcd
kube-system   service/kube-dns             ClusterIP   10.96.0.10       <none>        53/UDP,53/TCP                                    3d23h   k8s-app=kube-dns
kube-system   service/tiller-deploy        ClusterIP   10.108.27.198    <none>        44134/TCP                                        3d22h   app=helm,name=tiller
ric-e2        service/sctp                 NodePort    10.106.198.111   <none>        4488:30555/SCTP,2500:32416/SCTP,5566:30525/TCP   3d22h   app=ric-e2-ric-e2,release=ric-e2

**Kubeadm config file to creare the K8S cluster:**
apiVersion: kubeadm.k8s.io/v1alpha3
kubernetesVersion: v1.13.1
kind: ClusterConfiguration
apiServerExtraArgs:
  feature-gates: SCTPSupport=true
  #apiserver-advertise-address: 192.168.10.101
  #pod-network-cidr: 10.244.0.0/16
networking:
  dnsDomain: cluster.local
  podSubnet: 10.244.0.0/16
  #podSubnet: 192.168.0.0/16
  serviceSubnet: 10.96.0.0/12

apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
mode: ipvs

**Cilium yaml file taken from:**
v1.4.0-rc1.zip:
cilium-1.4.0-rc1/examples/kubernetes/1.13/cilium.yaml

Please note:
I was able to work with the K8S and SCTP service on Ubuntu 16.04 with the flannel network.
I tried with cilium also on the same platform and cilium version 1.2 but didn't work as well, so I upgrade to Ubuntu 18.04 with cilium version 1.4 as mentioned above.
 

Thanks,
Nati Malca

### PR DESCRIPTION
**Pull request by tgraf**
Todo:
 - [ ] CI tests

Fixes: #6907

Fixes https://github.com/cilium/cilium/issues/5719

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6912)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Review by borkmann — APPROVED** [review]
I think as-is SCTP multihoming might not work here as we'd need to parse the chunks and add all related IPs for this connection for policy enforcement.

**Comment by brb** [comment]
test-me-please

File: bpf/lib/csum.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 
 #define TCP_CSUM_OFF (offsetof(struct tcphdr, check))
 #define UDP_CSUM_OFF (offsetof(struct udphdr, check))
+#define SCTP_CSUM_OFF 8
```
Comment (joestringer) [inline bpf/lib/csum.h:28]: nit: `(offsetof(struct sctphdr, checksum))`

(From `linux/sctp.h`)

File: api/v1/models/proxy_statistics.go [hunk 2]
```diff
@@ -17,7 +17,7 @@ import (
 
 // ProxyStatistics Statistics of a set of proxy redirects for an endpoint
 // swagger:model ProxyStatistics
-// +k8s:deepcopy-gen=true
+
```
Comment (joestringer) [inline api/v1/models/proxy_statistics.go:20]: Is this deliberately removed? (Check for equivalent changes across all files)

**Review by joestringer — CHANGES_REQUESTED** [review]
Lots of seemingly unrelated deepcopy removals in this PR.

We should consider rejecting SCTP in the API with L7 rules, since there are likely no proxy implementations that will accept SCTP traffic.

How are we doing on verifier complexity in this PR?

Another TODO would be any documentation updates.

File: bpf/lib/csum.h (hunk 1 above)
Comment (tgraf) [reply bpf/lib/csum.h:28]: I looked in my `linux/sctp.h` and did not find that header. I assume it's only present in a later dev version. I'll get a copy from a recent kernel.

File: api/v1/models/proxy_statistics.go (hunk 2 above)
Comment (tgraf) [reply api/v1/models/proxy_statistics.go:20]: No, I'll fix it up. I've pushed the branch primarily to make sure the code is not lost.

**Comment by tgraf** [comment]
still alive

**Comment by tgraf** [comment]
Keep this alive

**Pull request closed without merging**